### PR TITLE
feat(protocol): l1 and l2 implementation of Lido integration

### PR DESCRIPTION
**Pull request by swarna1101**
This PR involves work regarding Taiko ~ Lido WstEth Bridge.

* This PR includes L1 contracts implementation, which are responsible for handling actions like deposit, L1 -> L2, communication, and withdrawals, bridge pausing mechanisms etc.
* Tests (incomplete)
* This PR also contains the L2 contracts implementation

### CODE REVIEW

**Comment by adaki2004** [comment]
First, i'd have several remarks:
- I'm not sure if this is the right place for the files. Both file-system level in the repo (packages/protocol), or if this is the suitable repo itself, given it runs under `GPL-3.0` license, while we have `MIT`. (Also pointed out by OpenZeppelin previously in their audit)
- I dont really get which is the code copied over and which is custom made by @swarna1101  ? I guess all of the solidity code is a take-over, in which case it is not at all final, since we need to adjust/tailor it to our Bridge.
- Would be good to have a concept, (a tiny/ugly diagram in excalidraw or paint would be enough) where the working mechanics of the whole bridging procedure is visible

The contracts need to be hooked into our Bridging system somehow. So "something" has to call the `Bridge`'s `sendMessage` function.

File: packages/protocol/contracts/thirdparty/lido/.gitmodules [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[submodule "lib/forge-std"]
+	path = lib/forge-std
+	url = https://github.com/foundry-rs/forge-std
```
Comment (dantaik) [inline packages/protocol/contracts/thirdparty/lido/.gitmodules:3]: we have modules defined in `packages/protocol`, do not make another sub-foundry project inside an existing one.

File: packages/protocol/contracts/thirdparty/lido/src/BridgeableTokens.sol [hunk 2]
```diff
@@ -0,0 +1,49 @@
+// SPDX-FileCopyrightText: 2022 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.20;
```
Comment (dantaik) [inline packages/protocol/contracts/thirdparty/lido/src/BridgeableTokens.sol:4]: May I know which files are copied over from other project as-is, which ones have our own modificaiton. 

If we copied files over, we need a URL to the original source file.

File: packages/protocol/contracts/thirdparty/lido/src/BridgeableTokens.sol [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+/// @author psirex
+/// @notice Contains the logic for validation of tokens used in the bridging process
+contract BridgeableTokens {
```
Comment (dantaik) [inline packages/protocol/contracts/thirdparty/lido/src/BridgeableTokens.sol:8]: abstract contract?

File: packages/protocol/contracts/thirdparty/lido/src/L1Executor.sol [hunk 4]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.20;
+
+import { IL1Executor } from "./interfaces/IL1Executor.sol";
+
+/// @dev Helper contract for contracts performing cross-domain communications
+contract L1Executor {
```
Comment (dantaik) [inline packages/protocol/contracts/thirdparty/lido/src/L1Executor.sol:8]: I don't see how these contracts interact with our native bridge...

File: packages/protocol/contracts/thirdparty/lido/src/BridgeableTokens.sol (hunk 2 above)
Comment (swarna1101) [reply packages/protocol/contracts/thirdparty/lido/src/BridgeableTokens.sol:4]: `BridgeableTokens.sol`, and `BridgingManager.sol` is taken from the already existing Lido implementation. Other file like `L1Executor.sol` is completely implemented by me.

**Comment by dantaik** [comment]
> First, i'd have several remarks:
> 
> * I'm not sure if this is the right place for the files. Both file-system level in the repo (packages/protocol), or if this is the suitable repo itself, given it runs under `GPL-3.0` license, while we have `MIT`. (Also pointed out by OpenZeppelin previously in their audit)
> * I dont really get which is the code copied over and which is custom made by @swarna1101  ? I guess all of the solidity code is a take-over, in which case it is not at all final, since we need to adjust/tailor it to our Bridge.
> * Would be good to have a concept, (a tiny/ugly diagram in excalidraw or paint would be enough) where the working mechanics of the whole bridging procedure is visible
> 
> The contracts need to be hooked into our Bridging system somehow. So "something" has to call the `Bridge`'s `sendMessage` function.

Given the GPL license, maybe adding a sub-project makes more sense. And we do need to mark our code to be MIT.

**Comment by dantaik** [comment]
Any update?

**Comment by swarna1101** [comment]
> Any update?

yes, I understood to make it compatible to our bridge, in the following commit, it will be:

- Compatible bridge
- Flow chart
- Minor Fixes with proper licensing

**Comment by dantaik** [comment]
> > Any update?
> 
> yes, I understood to make it compatible to our bridge, in the following commit, it will be:
> 
> * Compatible bridge
> * Flow chart
> * Minor Fixes with proper licensing

Ping me when this PR is ready. But I suggest we get this reviewed and approved by the end of this week.

**Comment by swarna1101** [comment]
> Any update?

I have updated the contract as per the Bridge Implementation and also, made sure it is compatible with our bridge.
e90cf757-b107-4d00-8ab5-64baa7e16166
@adaki2004 added a flow chart as well

**Comment by adaki2004** [comment]
It seems it is a custom Bridge architecture on the top of our Taiko native Bridge's cross-chain messaging - (which is good) - but i still lacking the bit of the context. For example i mean:
- Who implemented which contract exactly ? Which is a complete take-over, 100% (which might not require audit) and which is developped completely by you ?
- This is more of a component-kinda diagram than a flow chart. Would be cool to have a flow/sequence diagram to get to know the exact workflow (in order) to know what users and also reviews need to look and how.
- I see no unit tests - which would be very helpful also to get the context + flow a bit more.

If you could provide some answers to the first 2 points - that would definitely shorten my learning curve and can help a bit more.

**Comment by swarna1101** [comment]
> It seems it is a custom Bridge architecture on the top of our Taiko native Bridge's cross-chain messaging - (which is good) - but i still lacking the bit of the context. For example i mean:
> 
> * Who implemented which contract exactly ? Which is a complete take-over, 100% (which might not require audit) and which is developped completely by you ?
> * This is more of a component-kinda diagram than a flow chart. Would be cool to have a flow/sequence diagram to get to know the exact workflow (in order) to know what users and also reviews need to look and how.
> * I see no unit tests - which would be very helpful also to get the context + flow a bit more.
> 
> If you could provide some answers to the first 2 points - that would definitely shorten my learning curve and can help a bit more.

* All the contracts in `packages/protocol/contracts/thirdparty/lido` is taken completely from the existing Lido implementation, and all the contracts in `partners/lido/` is implemented by me.
* [this](https://cdn.discordapp.com/attachments/981277301100662804/1256229662686707742/Deposit_Flow.png?ex=6680028b&is=667eb10b&hm=4c3c950e041cee1a82118a7390a09f14c04274b58c3d4429ee714aac4a2b4d87&) is the deposit flow ,  [this](https://cdn.discordapp.com/attachments/981277301100662804/1256229663169314846/Withdraw_Flow.png?ex=6680028b&is=667eb10b&hm=e541db2337a51fcda19e5bdeb26923409ad51f8174fdf9865c2950474a3ec42b&) is Withdram Flow , and [this](https://cdn.discordapp.com/attachments/981277301100662804/1256230768401977499/FailMsg.png?ex=66800392&is=667eb212&hm=89824de6fe2b958e58901a95d8561f6298ee6f5c38615d934f4d4b153de891d7&) is fail msg . 
* Yes, unit tests, I will add in next

File: packages/protocol/contracts/thirdparty/lido/BridgeableTokens.sol [hunk 5]
```diff
@@ -0,0 +1,50 @@
+// SPDX-FileCopyrightText: 2022 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.24;
+
+/// @author psirex
+/// @notice Contains the logic for validation of tokens used in the bridging process
+/// @dev Taken from https://github.com/lidofinance/lido-l2/blob/main/contracts/BridgeableTokens.sol
+contract BridgeableTokens {
```
Comment (dantaik) [inline packages/protocol/contracts/thirdparty/lido/BridgeableTokens.sol:9]: So all these contracts are not upgradable?

File: packages/protocol/contracts/partners/lido/LidoL1Bridge.sol [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+        _;
+    }
+
+    /**
+     * @dev Modifier to ensure correct transfer of L1 tokens before and after a function call.
+     * @param isTo Boolean indicating whether tokens are being transferred to (`true`) or from
+     * (`false`) the contract.
+     * @param amount_ The amount of tokens being transferred.
+     */
+    modifier transferL1Tokens(bool isTo, uint256 amount_) {
```
Comment (dantaik) [inline packages/protocol/contracts/partners/lido/LidoL1Bridge.sol:57]: if the token is any arbitrary ERC20 token, then we need this diff-check, if we know it's a specific token implementaiton, we don't.

Is Lido token on L1 a regular ERC20 token?

File: packages/protocol/contracts/partners/lido/ILidoL1Bridge.sol [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        bytes calldata data_
+    )
+        external;
+
+    /**
+     * @notice Receives and processes a message from the L2 bridge
+     * @param _message The message received from the L2 bridge
+     * @param _proof The proof of the message
+     */
+    function receiveMessage(IBridge.Message calldata _message, bytes calldata _proof) external;
```
Comment (dantaik) [inline packages/protocol/contracts/partners/lido/ILidoL1Bridge.sol:89]: You have to implement the `IMessageInvocable` interface in IBridge.sol

Comment (dantaik) [reply packages/protocol/contracts/partners/lido/ILidoL1Bridge.sol:89]: You will also need to check permission inside the `onMessageInvocation` function, please checkout `checkProcessMessageContext` in ERC20Vault.

File: packages/protocol/contracts/partners/lido/ILidoL2Bridge.sol [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        uint256 amount_,
+        bytes calldata data_
+    )
+        external;
+
+    /**
+     * @notice Handles a failed message
+     * @param _message The failed message received
+     */
+    function handleFailMessage(IBridge.Message calldata _message) external;
```
Comment (dantaik) [inline packages/protocol/contracts/partners/lido/ILidoL2Bridge.sol:102]: You probably want to implement  `IRecallableSender.onMessageRecalled`.

File: packages/protocol/contracts/partners/lido/LidoL1Bridge.sol [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+            _message.from, l1Token_, l2Token_, from_, to_, amount_, data_
+        );
+    }
+
+    /**
+     * @notice Handles a failed message
+     * @param _message The failed message
+     * @param amount_to_receive The amount of tokens to be received as compensation
+     */
+    function handleFailMessage(
```
Comment (dantaik) [inline packages/protocol/contracts/partners/lido/LidoL1Bridge.sol:134]: this function must be `onMessageRecalled`

**Review by dantaik — CHANGES_REQUESTED** [review]

File: packages/protocol/contracts/partners/lido/LidoBridgedToken.sol [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+     * @param name_ The name of the ERC20 token
+     * @param symbol_ The symbol of the ERC20 token
+     * @param l2bridge_ The address of the L2 bridge
+     */
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        address l2bridge_
+    )
+        ERC20(name_, symbol_)
```
[the remark this answers is not in the document]
Comment (swarna1101) [reply packages/protocol/contracts/partners/lido/LidoBridgedToken.sol:32]: Lidos token decimals on L2 is 18.

File: packages/protocol/contracts/thirdparty/lido/BridgeableTokens.sol (hunk 5 above)
Comment (swarna1101) [reply packages/protocol/contracts/thirdparty/lido/BridgeableTokens.sol:9]: Currently not upgradable. Will make them upgradable

File: packages/protocol/contracts/partners/lido/LidoL1Bridge.sol (hunk 6 above)
Comment (swarna1101) [reply packages/protocol/contracts/partners/lido/LidoL1Bridge.sol:57]: Yes regular erc20. Will remove check

File: packages/protocol/contracts/partners/lido/ILidoL2Bridge.sol (hunk 8 above)
Comment (swarna1101) [reply packages/protocol/contracts/partners/lido/ILidoL2Bridge.sol]: used `basevault`, which inherits that.  if I separately inherit that, I should copy paste 2-3 functions from that basevault, so just reused

**Pull request closed without merging**